### PR TITLE
[AutoDiff upstream] Define `AutoDiffConfig`.

### PR DESCRIPTION
**Pull request by dan-zheng**
Define `AutoDiffConfig`: the configuration for a derivative function:
- Parameter indices.
- Result indices.
- Derivative generic signature (optional).

---

Progress towards TF-828: upstream `@differentiable` attribute type-checking.
`@differentiable` attribute type-checking mega-patch: https://github.com/apple/swift/pull/29091

---

Tests exist on `tensorflow` branch and will be upstreamed later.
Currently, not enough code has been upstreamed for meaningful testing.

### CODE REVIEW

File: include/swift/AST/AutoDiff.h [hunk 1]
```diff
@@ -148,10 +168,59 @@ void getSubsetParameterTypes(IndexSubset *indices, AnyFunctionType *type,
...
 using swift::AutoDiffDerivativeFunctionKind;
+using swift::GenericSignature;
+using swift::IndexSubset;
 
 template <typename T> struct DenseMapInfo;
 
+template <> struct DenseMapInfo<AutoDiffConfig> {
+  static AutoDiffConfig getEmptyKey() {
+    auto *ptr = llvm::DenseMapInfo<void *>::getEmptyKey();
+    // The `derivativeGenericSignature` component must be `nullptr` so that
```
Comment (dan-zheng) [inline include/swift/AST/AutoDiff.h]: Note: `GenericSignatureImpl::getCanonicalSignature()` crashes for `GenericSignature` with underlying `nullptr`, which seems like an anti-pattern. I'm not sure whether this is intentional, so I [posted a forum question](https://forums.swift.org/t/make-genericsignature-getcanonicalsignature-not-crash-for-nullptr-genericsignature/32409?u=dan-zheng).

Comment (dan-zheng) [reply include/swift/AST/AutoDiff.h]: @jckarter suggested defining `GenericSignature::getCanonicalSignature()` with the expected behavior as a fix. I'll investigate that in a follow-up.

**Comment by dan-zheng** [comment]
@swift-ci Please smoke test

File: include/swift/AST/AutoDiff.h (hunk 1 above)
Comment (dan-zheng) [reply include/swift/AST/AutoDiff.h]: `GenericSignature::getCanonicalSignature` done in https://github.com/apple/swift/pull/29105.

However, I'd like to defer rewriting uses of `GenericSignatureImpl::getCanonicalSignature` in `AutoDiffConfig` until changes have been downstreamed to and tested on `tensorflow` branch.

**Review by rxwei — APPROVED** [review]

**Pull request merged**